modules/standard/Sets.chpl

```diff
@@ -0,0 +1,395 @@
+      otherwise.
+
+      :arg x: The element to test for membership.
+      :return: Whether or not the given element is a member of this set.
+    */
+    proc const contains(const ref x: eltType): bool {
+      return false;
+    }
+
+    /*
+      Returns `true` if this set shares no elements in common with the set
+      `other`, and `false` otherwise.
+
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` are disjoint.
+    */
+    proc const isDisjoint(const ref other: set(eltType, ?)): bool {
+      return false;
+    }
+
+    /*
+      Returns `true` if this set and `other` have at least one element in
+      common.
+
+      :arg other: The set to compare against.
+      :return: Whether or not this set and `other` intersect.
+    */
+    proc const isIntersecting(const ref other: set(eltType, ?)): bool {
+      return false;
+    }
+
+    /*
+      Remove the item from this set with a value equal to `x`.
+
+      .. warning::
+
+        Removing an element from this set may invalidate existing references
+        to the elements contained in this set.
+    
+      :arg x: The element to remove.
+
+      :throws IllegalArgumentError: If the list contains no such element.
+    */
+    proc remove(const ref x: eltType): eltType throws {
+      return;
+    }
+
+    /*
+      Remove the item `x` if it is present in this set, but do nothing if it
+      is not.
+
+      .. warning::
+
+        Discarding an element from this set may invalidate existing references
+        to the elements contained in this set.
+
+      :arg x: The element to discard.
+    */
+    proc discard(const ref x: eltType) {
+      return;
+    }
+
+    /*
+      Clear the contents of this set.
+
+      .. warning::
+
+        Clearing the contents of this set will invalidate all existing
+        references to the elements contained in this set.
+    */
+    proc clear() {
+      return;
+    }
+
+    /*
+      Iterate over the elements of this set.
+
+      :yields: A reference to one of the elements contained in this set.
+    */
+    iter these() ref {
+      return nil;
+    }
+
+    /*
+      Write the contents of this set to a channel.
+
+      :arg ch: A channel to write to.
+    */
+    proc const writeThis(ch: channel) {
+      return;
+    }
+
+    /*
+      Returns `true` if this set contains zero elements.
+
+      :return: `true` if this set is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+      return false;
+    }
+
+    /*
+      The current number of elements contained in this set.
+    */
+    inline proc const size {
+      return 0;
+    }
+
+    /*
+      Returns a new DefaultRectangular array containing a copy of each of the
+      elements contained in this set. The returned array is sorted according
+      to the default sort order of its elements.
+    */
+    proc const toArray(): [] eltType {
```

Comment:
I'd make the sorting optional, or make the order unspecified and let the caller sort the resulting array as desired.  Or allow the user to provide their own `comparator`.  Unless you're expecting to be able to provide that particular sorted order efficiently?

Reply:
Good point about the comparator. Also it's important to note that not all types are currently sortable.

One reasonable strategy would be to allow a comparator argument, e.g. with `proc const toArray(comparator:? = none)`. The body of this function can check if comparator is not `none` and sort in that event. I like this adjustment to the strategy because:
 * it makes a common use case (wanting to sort when creating an array) easy
 * it still allows for flexibility

I am not expecting that the sort is any more efficient than if the user sorted the result of toArray themselves.